src/Raven.Embedded/RavenEmbedded.cs

```diff
@@ -10,137 +10,138 @@
 using Raven.Client.ServerWide;
 using Raven.Client.ServerWide.Operations;
 using Sparrow.Logging;
-
+using Raven.Client.Util;
 
 namespace Raven.Embedded
 {
-    public class RavenEmbedded
+    public class EmbeddedServer :  IDisposable
     {
-        private Process _ravenDbSlavedProcess;
-        private readonly Logger Logger = LoggingSource.Instance.GetLogger<RavenEmbedded>("Embedded");
-        private Task<Uri> _serverTask;
+        public static EmbeddedServer Instance = new EmbeddedServer();
+
+        private readonly Logger Logger = LoggingSource.Instance.GetLogger<EmbeddedServer>("Embedded");
+        private Lazy<Task<(Uri ServerUrl, Process ServerProcess)>> _serverTask;
         private long _initialized;
 
-        private readonly ConcurrentDictionary<string, IDocumentStore> _documentStores = new ConcurrentDictionary<string, IDocumentStore>();
+        private readonly ConcurrentDictionary<string, Lazy<Task<IDocumentStore>>> _documentStores = new ConcurrentDictionary<string, Lazy<Task<IDocumentStore>>>();
 
         public void StartServer(ServerOptions options = null)
         {
-            // Start the server
+            options = options ?? ServerOptions.Default;
 
-            if (Interlocked.CompareExchange(ref _initialized, 1, 0) != 0)
-                throw new InvalidOperationException("Already have been initialized");
+            var startServer = new Lazy<Task<(Uri ServerUrl, Process ServerProcess)>>(() => RunServer(options));
+            if (Interlocked.CompareExchange(ref _serverTask, startServer, null) != null)
+                throw new InvalidOperationException("The server was already started");
 
-            options = options ?? ServerOptions.Default;
-            _serverTask = RunServer(options);
+            // this forces the server to start running in an async manner.
+            GC.KeepAlive(startServer.Value);
         }
 
-        public static RavenEmbedded Instance = new RavenEmbedded();
 
-        public async Task<IDocumentStore> GetDocumentStore(string database)
+        public IDocumentStore GetDocumentStore(string database)
         {
-            if (Interlocked.Read(ref _initialized) == 0)
-                throw new InvalidOperationException("Please run StartServer before trying to get document store");
+            return AsyncHelpers.RunSync(() => GetDocumentStoreAsync(database));
+        }
 
-            if (string.IsNullOrEmpty(database))
-                throw new ArgumentException(nameof(database));
+        public IDocumentStore GetDocumentStore(GetDocumentStoreOptions options)
+        {
+            return AsyncHelpers.RunSync(() => GetDocumentStoreAsync(options));
+        }
 
-            if (Logger.IsInfoEnabled)
-                Logger.Info($"GetDocumentStore for db ${ database }.");
+        public Task<IDocumentStore> GetDocumentStoreAsync(string database)
```

Comment:
probably we should consider adding cancellation token here

...and to all public (but applicable) methods.